### PR TITLE
feat(aws-rds): proxy endpoints

### PR DESCRIPTION
**Pull request by Jacco**
fixes: #14186
----

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [x] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/master/INTEGRATION_TESTS.md)?
	* [x] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-rds/lib/proxy.ts [hunk 1]
```diff
@@ -341,6 +348,11 @@ abstract class DatabaseProxyBase extends cdk.Resource implements IDatabaseProxy
   public abstract readonly dbProxyArn: string;
   public abstract readonly endpoint: string;
 
+  /**
+   * Collection of all the proxy endpoints of this proxy
+   */
+  public readonly endpoints: string[] = [];
```
Comment (corymhall) [inline packages/@aws-cdk/aws-rds/lib/proxy.ts:354]: I'm not sure if this adds any value. If you have added a bunch of endpoints you would probably want to know which _specific_ one you want to connect to.

File: packages/@aws-cdk/aws-rds/lib/proxy.ts [hunk 2]
```diff
@@ -524,3 +546,219 @@ function toConnectionPoolConfigurationInfo(
...
+/**
+ * Options for a new DatabaseProxyEndpoint
+ */
+export interface DatabaseProxyEndpointOptions {
+  /**
+   * The identifier of the DB proxy endpoint.
+   *
+   *  @default - Generated by CloudFormation (recommended)
+   */
+  readonly dbProxyEndpointName: string;
```
Comment (corymhall) [inline packages/@aws-cdk/aws-rds/lib/proxy.ts:574]:
```suggestion
  readonly dbProxyEndpointName?: string;
```

Also change the `@default` to be "Generated by CDK"

File: packages/@aws-cdk/aws-rds/lib/proxy.ts [hunk 3]
```diff
@@ -524,3 +546,219 @@ function toConnectionPoolConfigurationInfo(
...
+
+  /**
+   * Access to network connections.
+   */
+  public readonly connections: ec2.Connections;
+
+  private readonly resource: CfnDBProxyEndpoint;
+
+  constructor(scope: Construct, id: string, props: DatabaseProxyEndpointProps) {
+    super(scope, id, { physicalName: props.dbProxyEndpointName || id });
```
Comment (corymhall) [inline packages/@aws-cdk/aws-rds/lib/proxy.ts:737]: I know the proxy resource does it like this, but can we do this instead?

```ts
physicalName: props.dbProxyEndpoint || Lazy.string({ produce: () => Names.uniqueId(this) }),
``

**Review by corymhall — DISMISSED** [review]
First pass comments. Also need to add:

1. README updates
2. integration test
3. Review/remove the TODO comments.

File: packages/@aws-cdk/aws-rds/lib/proxy.ts (hunk 1 above)
Comment (Jacco) [reply packages/@aws-cdk/aws-rds/lib/proxy.ts:354]: True

File: packages/@aws-cdk/aws-rds/lib/proxy.ts (hunk 2 above)
Comment (Jacco) [reply packages/@aws-cdk/aws-rds/lib/proxy.ts:574]: Good one

File: packages/@aws-cdk/aws-rds/lib/proxy.ts [hunk 4]
```diff
@@ -524,3 +546,219 @@ function toConnectionPoolConfigurationInfo(
     sessionPinningFilters: props.sessionPinningFilters?.map(_ => _.filterName),
   };
 }
+
+/**
+ * The type of storage.
```
[the remark this answers is not in the document]
Comment (Jacco) [reply packages/@aws-cdk/aws-rds/lib/proxy.ts]: Oops

**Comment by Jacco** [comment]
@corymhall think the PR is ready for a 2nd round. A few things are still to be discussed:

- isDefault
- and the scope parameter

Also the integ test is not a really working example beause the 2 vpc's involved are not configured to be able to talk to eacg other (I think).

Maybe the documentation needs some elaboration too.

**Review by corymhall — CHANGES_REQUESTED** [review]
Another round of review. Some questions that I have after looking more into this that I think we should account for.

1. It's not exactly clear how the "default" endpoint works. Is that the endpoint that you get from creating a DbProxy? If so, why does the `isDefault` parameter even exist? Is it because you can _update_ the "default" endpoint?
2. `READ_ONLY` only works with Aurora DB clusters. Can we add some sort of validation to make sure that users don't try to add `READ_ONLY` endpoint to a non  Aurora cluster?

There is some good documentation on `TargetRole` [here](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/rds-proxy-endpoints.html#rds-proxy-endpoints.CreatingEndpoint) that I think we should include.

**Pull request closed without merging**